file_util/src/file.cc

```diff
@@ -0,0 +1,142 @@
+#include <leatherman/file_util/file.hpp>
+#include <boost/nowide/fstream.hpp>
+#include <boost/nowide/cenv.hpp>
+#include <boost/filesystem/operations.hpp>
+#include <boost/system/error_code.hpp>
+#include <sstream>
+#include <leatherman/logging/logging.hpp>
+
+using namespace std;
+
+namespace leatherman { namespace file_util {
+
+namespace boost_error = boost::system::errc;
+namespace boost_file = boost::filesystem;
+
+    bool each_line(string const& path, function<bool(string&)> callback)
+    {
+        boost::nowide::ifstream in(path.c_str());
+        if (!in) {
+            return false;
+        }
+
+        string line;
+        while (getline(in, line)) {
+            if (!callback(line)) {
+                break;
+            }
+        }
+        return true;
+    }
+
+    string read(string const& path)
+    {
+        string contents;
+        if (!read(path, contents)) {
+            return {};
+        }
+        return contents;
+    }
+
+    bool read(string const& path, string& contents)
+    {
+        boost::nowide::ifstream in(path.c_str(), ios::in | ios::binary);
+        ostringstream buffer;
+        if (!in) {
+            return false;
+        }
+        buffer << in.rdbuf();
+        contents = buffer.str();
+        return true;
+    }
+
+    bool file_readable(const std::string &file_path) {
+        bool exists { false };
+        if (file_path.empty()) {
+            LOG_WARNING("file path is an empty string");
+        } else{
+            boost::system::error_code ec;
+            boost_file::file_status status = boost_file::status(file_path.c_str(), ec);
+            if (boost_file::status_known(status)) {
```

Comment:
Is that sufficient to check whether it's a file (and not a directory)? I think we also need to check `exists` and `!is_directory`.